### ISSUE TITLE
Code Gen'd fast path CBOR serialization/deserialization

### ISSUE THREAD

**Issue by hannahhoward** [ipld/go-ipld-prime#141]
Hello!

With codegen getting pretty advanced, I'd like to understand where generating fast path CBOR serialization/deserialization sits on the priority lists. I communicated some of my own findings a while back to @warpfork about speeds with codegen'd nodes vs native go structs + @whyrusleeping 's [cbor-gen](https://github.com/whyrusleeping/cbor-gen). (see https://github.com/ipld/go-ipld-prime/issues/83 for a summary)

We get pretty close to native structs + cbor-gen with code gen'd nodes, but we still pay an overhead cost for using refmt for cbor serialization/deserialization. I'm wondering what it would take, and what the interest level is, in either porting cbor-gen code over to the code gen in go-ipld-prime, or writing go-ipld-prime code from scratch (honestly I imagine some mix using cbor-gen as a reference).

**Comment by warpfork** [comment]
**groups of potential work**:

- Work grouping 1: if we're really close already, just by getting good at avoiding allocations through the design of NodeAssemblers, I think I'm satisfied with that.  We should value any further improvements to this highly, because any payoffs apply to all codecs, even ones we might not have even invented yet.  (But I'm not sure how much room there is still to go, here.)

- Work grouping 2: I think I'd like to rewrite some of the cbor stuff directly in go-ipld-prime (and kinda ditch refmt's indirections) at this point for *various* reasons.  But I'm not sure how much performance there is to buy here.  I estimate somewhere between 5% and 10%ish.  This costs development time, but not much else (wouldn't be expecting bigger codegen sizes, etc; most of the abstractions land on the same edges).

- Work grouping 3: It's possible to go extremely far and _generate **additional** code_ which eschews _NodeAssemblers_ entirely, and parses CBOR directly into the target structures.  This will be speed maximizing, but takes a _significant_ amount of development work, and results in _bigger codegen outputs_, and is _not reusable effort_ between codecs, and is also _useless for untyped data_.

**present opinions on what's most valuable**:

I don't know if there's much room for improvement in group 1, so there's a priority of NaN there.

The room for improvement in group 2 is "some" but not stellar.  I'd say it's probably "medium" priority.  (At the moment, that means "a lot less important than other features such as getting Schema DSL parsers connected to codegen, and making it a cleaner CLI tool".)

The room for improvement in group 3 might be sizable, or it might not, and we really don't know until we build it.  But it's also making a *lot* of tradeoffs and is very nonreusable.  So I'm not inclined to prioritize it much at all unless someone wants to contribute that effort because it's super important to them.

If I'm extracting the numbers from #83 correctly (`4776305ns/5102731ns`), only we're roughly 6 or 7% behind cbor-gen's performance already.  It also looks like the handroll custom CBOR you made for science only gained another 1% in time (although it saved a noticeable amount of garbage), so... I haven't looked into the details of that much, but it might suggest that work in grouping 3, especially considering all its costs and impact limitations, is really not worth it unless we've got experimental time to spare.

We've also probably got some low-hanging fruit to get from doing memory pooling for the serialization machinery, and that's likely to happen as part of work grouping 2.  (Or sooner: there's some API refactors planned involving a LinkSystem concept that might prompt this very presently, possibly in the next few weeks.)  How much effect that has might depend on the workload patterns, though.

**tl;dr**:

A small amount of work on the current CBOR codec is probably in the near future.  A larger piece of work on it seems valuable but doesn't have a date attached.  A CBOR-specific codegen extension currently looks like a dubious investment of time, unless I'm either misinterpreting something or new information appears, but I'm also certainly not opposed to it if someone else wants to bet otherwise and invest the time.

**Comment by rvagg** [comment]
Can we have some of those pretty Go profile charts that show were time is being spent in this pipeline? That might be really interesting and point to low hanging fruit over in refmt that we can shave.
